File: problem_3/problem_3.php
<?php

function faster_largest_prime($n) {
    $factor = 2;
    $last_factor = 1;
    while ($n > 1) {
        if ($n % $factor == 0) {
            $last_factor = $factor;
            $n = $n / $factor;
        } else {
            $factor++;
        }
    }
    return $last_factor;
}

var_dump(faster_largest_prime(13195));
var_dump(faster_largest_prime(600851475143));

?>

File: problem_3/problem_3.rs
// compile with rustc problem_3.rs
// run as ./problem_3

fn main() {
    println!("{}", faster_largest_prime(13195));
    println!("{}", faster_largest_prime(600851475143));
}

fn faster_largest_prime(mut n: i64) -> i64 {
    let mut factor = 2;
    let mut last_factor = 1;
    loop {
        if n == 1 {
            return last_factor;
        }
        if n % factor == 0 {
            last_factor = factor;
            n = n / factor;
        } else {
            factor += 1;
        }
    }
}